### PR TITLE
Feature/2132/terra

### PR DESCRIPTION
**Pull request by coverbeck**
Add launch with Terra button

ga4gh/dockstore#2132

Restructured code to make it easier to add more buttons in the future; it is now mainly just adding the HTML in launch-third-party.component.html, although a new logo will also need to be added, and dockstore.model.ts should be updated with the launch with url. See comments in launch-third-party.component.ts.

Here's what it looks like. Thoughts on the green Terra logo? Should we try to get an all-white version of the logo so it fits in better with the other buttons?

Screenshot from 2019-03-10 15-49-30

Also, if this PR gets approved, I will create a ticket to move launch-third-party/state to a higher level. The new WDL viewer is also fetching primary and secondary descriptors, so that component and this one should use the same state. But I wanted to limit the scope of the change for now.

### CODE REVIEW

File: src/app/workflow/launch-third-party/launch-third-party.component.ts [hunk 1]
```diff
@@ -1,70 +1,150 @@
...
+import { HttpUrlEncodingCodec } from '@angular/common/http';
+import FileTypeEnum = ToolFile.FileTypeEnum;
 
-const importHttpRegEx: RegExp = new RegExp(/^\s*import\s+"https?/, 'm');
+/**
+ *  A component that has buttons linking to external platforms that can launch workflows hosted on Dockstore.
+ *
+ *  ## Instructions for adding a new button
+ *
+ *  1. Update dockstore.model.ts with a new property that has the url of the external platform.
```
Comment (denis-yuen) [inline src/app/workflow/launch-third-party/launch-third-party.component.ts]: To be tedious, also update the default URL to compose setup

Comment (coverbeck) [reply src/app/workflow/launch-third-party/launch-third-party.component.ts]: Reminds me that I need to submit a PR myself for compose_setup.

File: src/app/workflow/launch-third-party/launch-third-party.component.ts [hunk 2]
```diff
@@ -1,70 +1,150 @@
...
-  fireCloudURL: string;
-  dnanexusURL: string;
-  wdlHasHttpImports: boolean;
-  wdlHasFileImports: boolean;
-  wdlHasContent: boolean;
-  isWdl: boolean;
+  /**
+   * Indicates whether the selected version has any content
+   */
+  hasContent$ = this.descriptorsQuery.hasContent$;
```
[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/workflow/launch-third-party/launch-third-party.component.ts]: It is being called exactly once for each time it is in the HTML -- 8 times.

Using share reduces it to one call. Updating in next commit

File: src/app/workflow/launch-third-party/launch-third-party.component.ts [hunk 3]
```diff
@@ -1,70 +1,150 @@
...
+      this.trsUrlAsQueryValue = new HttpUrlEncodingCodec()
+        .encodeValue(this.trsUrl);
+      this.workflowPathAsQueryValue = new HttpUrlEncodingCodec().encodeValue(this.workflow.full_workflow_path);
     }
   }
 
+  ngOnDestroy(): void {
+    this.ngUnsubscribe.next();
+    this.ngUnsubscribe.complete();
+    this.descriptorsQuery.destroy();
```
[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/workflow/launch-third-party/launch-third-party.component.ts]: Removed this whole method in combination with extending Base.

**Pull request merged**